### PR TITLE
Lynn Trickeyfinished homework questions

### PR DESCRIPTION
**Pull request by ltrickey**
# Recursion Tracing

Thanks for doing some brain yoga. You are now submitting this assignment!

## Comprehension Questions
|  Question 	|  Answer 	|
|------------	| ---------	|
|  Did you define all the recursion terms?	|   Yup.	|
|  How do you feel about tracing recursive problems? (good, bad, ugly, neutral)	|   Pretty good!	|
|  Would you like help/follow up about recursion?	|   I'd love to know more about common recursion problems - ie the palindrome. 	|

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -39,10 +39,17 @@ def mystery2(n)
...
+- What is mystery2(9005)? 14
+- What is mystery2(-123)? -123
 - _Added Fun: How could we make `mystery2(-123)` work the way we might expect it to work instead of the way it does?_
+- It looks like this method is created in order to add all the digits together of each number.  Meaning a number entered 123 should result in 6.  To account for negative numbers, we could create an earlier if statement:
+
+if n < 0 && n >= -10
+  return n
+else
+  return (n%10) - mystery2(n/10)
+end
```
Comment (sudocrystal) [inline README.md:52]: Hmm... I don't think this is right. What happens if you did mystery2(-12)? I would expect -3, but I don't think that's what your modification would yield.

File: README.md [hunk 2]
```diff
@@ -75,9 +84,11 @@ def mystery4(b,e)
 end
 ```
 
-- What is mystery4(10,2)?
-- What is mystery4(4,3)?
-- What is mystery4(5,0)?
+- What is mystery4(10,2)? 100
+- What is mystery4(4,3)? 64
+- What is mystery4(5,0)? 5
```
Comment (sudocrystal) [inline README.md:89]: Double check mystery4(5,0).

File: README.md [hunk 3]
```diff
@@ -90,10 +101,13 @@ def mystery5(s)
...
-- What is mystery5("hi")?
-- What is mystery5("")?
-- What is mystery5("Hi, there!")?
+ What is mystery5("hi")? "**" **
+- What is mystery5("")? ""
+- What is mystery5("Hi, there!")? = "**********"
 - _Added Fun: How could we make only alphabetic characters to be changed to stars?_
+This method is creating a string of stars * that replace every character in the string.
+
+You could add a statement: if s[0].include?(a..z) or [a-zA-Z] before running the return statement to make sure the space has a letter.  Otherwise you could have it return a space " " + mystery5(s[1..-1]).  Although this is using ruby magic and I'm not sure how to do it without.  I'd love to know!
```
Comment (sudocrystal) [inline README.md:111]: This looks like a good solution to me. Other languages have similar things that allow you to determine if something is a letter or not, so I wouldn't consider this ruby magic. I don't know how you'd know that until you've been exposed to other languages more though.

One question to think about is would you just strip out all none alphabetic characters? Or would you leave them in place? i.e. would the rest of mystery5modifed("Hi, there!") result in "*******" or "**, *****!" I was thinking the latter, but looks like you might have been thinking the former.

File: README.md [hunk 4]
```diff
@@ -110,7 +124,21 @@ def mystery6(s)
...
-- What is mystery6("Hi, there!")?
+- What is mystery6("goodnight moon")? " moon goodnight"
+- What is mystery6("Ada Developers Academy")? " Academy Developers Ada"
+- What is mystery6("Hi, there!")? " there Hi"
+
+This method reverses the order of words.
+
 - _Added Fun: How could we make the reversal happen by letter, instead of by word (i.e. Make it so that mystery6("goodnight moon") returned "noom thgindoog")?_
+
+At first this seemed impossible but I realized it was actually simpler than the first recursive method!  Cool!  You could rewrite the method this way, and it will reverse the string by letter instead of by word:
```
Comment (sudocrystal) [inline README.md:135]: :)

File: README.md [hunk 5]
```diff
@@ -110,7 +124,21 @@ def mystery6(s)
...
+```
+def mystery6(s)
+  if s == nil || s.length == 0
+    return ""
+  else
+    return mystery6(s[1..-1]) + s[0]
+  end
+end
+```
```
Comment (sudocrystal) [inline README.md:144]: right on!

**Comment by sudocrystal** [comment]
# Recursion Tracing

Nice job, especially on trying those added fun pieces!

File: README.md (hunk 1 above)
Comment (ltrickey) [reply README.md:52]: @sudocrystal I have tried to mess around with this, after remembering that mod doesn't work well with negative numbers.  I remember in class that you did something to retain whether the number was negative or positive, then multiply the result by that before returning.  I can't seem to get that to work (not sure where to put it with the multiple returns).  Do you have a code snipped I could look at or a suggestion?  Thank you!

File: README.md (hunk 2 above)
Comment (ltrickey) [reply README.md:89]: Duh.  This should be 1, not 5.

**Comment by sudocrystal** [comment]
Lines 3-7 of my solution are how I remembered the negative number in class - but I’m not sure that’s what you want here: https://github.com/sudocrystal/even-digits/blob/solutions/lib/solution-iterative.rb <https://github.com/sudocrystal/even-digits/blob/solutions/lib/solution-iterative.rb>

In this case, I think you want to utilize and absolute value function.

-- 
Crystal Hess
Program Director, Ada Developers Academy
http://adadevelopersacademy.org/ <http://adadevelopersacademy.org/> | @adaacademy | 425.298.7549 (cell)

**Comment by ltrickey** [comment]
Ahh, that makes more sense.  Thank you!

**Pull request closed without merging**